keps/sig-storage/661-statefulset-volume-resize/README.md

```diff
@@ -0,0 +1,1112 @@
+  * Another reason is that we want for the PVC reconciliation to match the cadence of pod creation/modification in the StatefulSet. If the user uses monotonic (`OrderedReadyPodManagement`) roll out strategy, we'd want the PVC reconciliation process to respect that strategy. This is especially important considering that PVC reconciliation could fail asynchronically (eg, due to out of space/exceeding quota issues) and we want the failure to be predictable.
+  * Another reason is that some CSI drivers don't support offline expansion (meaning, the volume has to be published/attached to a pod)
+  * This means that if the StatefulSet is scaled to 0, and the user resizes one of the PVC templates, the change won't be reconciled (even if the PVC objects exist), until the StatefulSet is scaled up again.
+* The reason for only waiting for the replicas/pods to be in a running state, and not necessarily in a ready state, is that the user might have readiness checks that will fail until the PVC is resized. Waiting for the pod to become fully ready, might put us in a deadlock.
+  * For example, imagine a readiness check that tests that the filesystem has at least 10% space left.
+
+On top of it, when we ecnounter an error during patching the PVC 
+  * If the error is transient (not part of [Errors When Patching The PVC Object](#when-patching-the-pvc-object)), record it on the StatefulSet object, and return the error immediately (this will stop the reconciliation process from progressing to other replicas and will trigger a retry).
+  * If the error is not transient (part of [Errors When Patching The PVC Object](#when-patching-the-pvc-object)), record it on the StatefulSet object, and
+    * If the roll out strategy of the StatefulSet is monotonic (`OrderedReadyPodManagement`), return immediately - but do not return the error, because if the error is not transient, we do not want to retry.
+    * If the roll out strategy is burst (not monotonic), continue.
+
+```
+for pvcTemplate in statefulset {
+  statefulset.status.pvcTemplateStatus[pvcTemplate.name].readyReplicas = 0
+}
+
+for replica in replicas {
+  // this is part of the existing replicas loop in the StatefulSet reconciliation process (https://github.com/kubernetes/kubernetes/blob/d89d5ab2680bc74fe4487ad71e514f4e0812d9ce/pkg/controller/statefulset/stateful_set_control.go#L397)
+  // ...
+  if getPodRevision(replica) == statefulset.revision && replica.Status.Phase == PodRunning && replica.DeletionTimestamp == nil {
+    for pvcTemplate in statefulset {
+      pvc, err = getPodPVC(replica, pvcTemplate)
+      if err != nil {
+        // this is either a transient error, or an error that shouldn't happen, because the pod is running and the PVC should exist
+        recordError(statefulset, err)
+        return err
+      }
+
+      if requestStorage(pvcTemplate) != requestStorage(pvc) {
+        err = patchPVC(pvc, requestStorage(pvcTemplate))
```

Comment:
How can users ever recover from a failed resize when the controller keeps overwriting PVC.spec capacity?

For example, the controller has successfully resized 2 PVCs out of 5, but the third resize is stuck on non-transient error (e.g. no space left in the storage backend). User can't reduce volumeClaimTemplate size, because some PVCs were already resized and at the same time they can't reduce size of PVC-3, because the controller will keep overwriting it.

Reply:
I'm not sure that this is an issue with the reconciliation logic you're pointing to. All the PVCs in the statefulset must eventually reconcile to the same size. You can't partially recover from partial failures

I make the agument here https://github.com/kubernetes/enhancements/blob/95a7fe96a22b2c3ba467df2be9879439d23c5d3f/keps/sig-storage/661-statefulset-volume-resize/README.md#asynchronous-errorsfailures 

> In such case, the user won't be able to apply the mechanisms in [KEP 1790](https://github.com/kubernetes/enhancements/blob/95a7fe96a22b2c3ba467df2be9879439d23c5d3f/keps/sig-storage/1790-recover-resize-failure/README.md) and their only recourse would be to request more quota from the provider, or downsize their StatefulSet (number of replicas).

Your only recourse would be increasing the backend storage quota or downsizing the statefulset replicas

Reply:
I don't like this approach. There should be a way how to recover from a partially failed resize, otherwise Kubernetes will re-try resizing volumes forever. Scaling down a StatefulSet or adding more storage to the backend are valid, still, it may not be enough in all cases.

What if StatefulSet admission allowed decreasing volumeClaimTemplate size if some PVCs are not resized yet (*)? Then the StatefulSet controller would shrink PVCs that it can shink, so the resizer can try with a smaller size or even stop resizing if the new size is equal to the current size in the storage backend. In the worst case, user ends up with the same PVs as with the partially resized StatefulSet, it just stops re-trying. Some PVCs will be bigger than the current PVC template size. IMO it could be fixed by some status field that shows total allocated storage capacity, so users can see the difference.

*) I know it's racy to validate a StatefulSet against PVCs that are being resized, it's more like best effort. The real logic must be in the StatefulSet controller and it must recover from errors when shrinking PVC.spec.requests.

Reply:
> Kubernetes will re-try resizing volumes forever

That's correct - but note that it will not block the StatefulSet controller, the user will still be able to make changes to pod spec.

In the current design, if the StatefulSet controller fails to patch the PVC object (due to the requested size being lower than the capacity), it will actually block the StatefulSet controller (unless it's a burst roll out) until the error is addressed by the user. 
Are you suggesting relaxing this (ie if patching the PVC object results in an error due to trying to downsize the PVCs that have been successfully resized, ignore it, or at least, don't make it block the StatefulSetController)?

In normal circumstances (everything is fully reconciled, no failure/partial-failure), it means that if the user tries to decrease the size of a PVC template, we'll let them do it